File: src/main/kotlin/dev/hawu/plugins/queuedsudo/WorldGroup.kt
package dev.hawu.plugins.queuedsudo

import dev.hawu.plugins.api.utils.Strings.toUUID
import dev.hawu.plugins.queuedsudo.executables.Executable
import org.bukkit.OfflinePlayer
import org.bukkit.configuration.serialization.ConfigurationSerializable
import org.bukkit.entity.Player
import java.util.*

class WorldGroup(
    val uuid: UUID,
    var name: String? = null,
    val worlds: MutableList<String> = mutableListOf(),
    val executables: MutableMap<UUID, MutableList<Executable>> = mutableMapOf(),
) : ConfigurationSerializable {
    
    fun queueExecutable(player: OfflinePlayer, executable: Executable) {
        if(player.isOnline && player.player.world.name in worlds) return executable.run(player.uniqueId)
        executables.putIfAbsent(player.uniqueId, mutableListOf())
        executables[player.uniqueId]!!.add(executable)
    }
    
    fun runAllExecutables(player: Player) {
        val list = executables[player.uniqueId] ?: return
        list.forEach {
            it.run(player.uniqueId)
        }
        executables.remove(player.uniqueId)
    }
    
    override fun serialize() = mapOf(
        "uuid" to uuid.toString(),
        "name" to name,
        "worlds" to worlds,
        "executables" to executables.mapKeys { it.key.toString() },
    )
    
    override fun equals(other: Any?) = other is WorldGroup && other.uuid == uuid
    override fun toString() = "WorldGroup{uuid=$uuid}"
    override fun hashCode() = uuid.hashCode()
    
    companion object {
        
        @Suppress("UNCHECKED_CAST")
        @JvmStatic
        fun deserialize(map: Map<String, Any>) = WorldGroup(
            uuid = map["uuid"].toString().toUUID(),
            name = map["name"]?.toString(),
            worlds = map["worlds"] as MutableList<String>,
            executables = map["executables"] as MutableMap<UUID, MutableList<Executable>>,
        )
        
    }
    
}

File: src/main/scala/dev/hawu/plugins/queuedsudo/WorldGroup.scala
package dev.hawu.plugins.queuedsudo

import dev.hawu.plugins.queuedsudo.executables.Executable
import org.bukkit.Material
import org.bukkit.configuration.serialization.ConfigurationSerializable

import java.util
import java.util.concurrent.atomic.AtomicInteger
import scala.collection.mutable
import scala.jdk.CollectionConverters.*

/**
 * Represents a group of worlds.
 *
 * @param name   the name of the group
 * @param worlds the worlds in the group
 */
case class WorldGroup(
   var name: String,
   var icon: Material = Material.CHEST,
   worlds: mutable.Set[String] = mutable.Set[String](),
) extends ConfigurationSerializable :

   /**
    * Serializes the world group into a data map.
    *
    * @return the data map
    */
   override def serialize(): util.Map[String, Any] = mutable.Map(
      "name" -> name,
      "icon" -> icon.name,
      "worlds" -> worlds.toList.asJava
   ).asJava

   override def hashCode(): Int = name.hashCode

   override def equals(obj: Any): Boolean = obj match
      case that: WorldGroup => that.name == this.name
      case _ => false

   override def toString: String = s"WorldGroup(name=$name,icon=${icon.name},worlds=$worlds)"

/**
 * Companion object for [[WorldGroup]].
 */
object WorldGroup:

   /**
    * Deserializes the data map into a world group.
    *
    * @param map the data map
    * @return the world group
    */
   def deserialize(map: util.Map[String, Any]): WorldGroup =
      val name = map.get("name").asInstanceOf[String]
      val icon = Some(Material.valueOf(map.get("icon").toString)).getOrElse(Material.CHEST)
      val worlds = new util.HashSet(map.get("worlds").asInstanceOf[util.List[String]]).asScala
      WorldGroup(name, icon, worlds)
